File: agenticpdf-rs/apps/reader/android/app/src/main/java/ai/nervosys/apdf/PageView.kt
// SPDX-License-Identifier: AGPL-3.0-or-later
package ai.nervosys.apdf

import android.content.Context
import android.graphics.Bitmap
import android.graphics.Canvas
import android.graphics.Color
import android.graphics.Paint
import android.graphics.Path
import android.graphics.RectF
import android.util.AttributeSet
import android.util.Base64
import android.view.GestureDetector
import android.view.MotionEvent
import android.view.View
import org.json.JSONArray
import org.json.JSONObject
import kotlin.math.abs

/** How many decoded masks to hold before starting again. */
private const val MAX_GLYPH_TILES = 4096

/**
 * Draws a page by replaying the recording the Rust core produced.
 *
 * The page-painting decisions — where each glyph sits, which paths fill, how
 * the page is fitted and flipped — were all made in shared Rust. What is left
 * here is a switch over primitives, which is why this file is short and why
 * Android cannot drift from the desktop and web renderings.
 */
class PageView @JvmOverloads constructor(
    context: Context,
    attrs: AttributeSet? = null,
) : View(context, attrs) {

    var zoom: Float = 1f
        set(value) {
            field = value.coerceIn(0.1f, 8f)
            invalidate()
        }

    /** Called when a horizontal swipe should turn the page. */
    var onSwipePage: ((Int) -> Unit)? = null

    /** Set when the document has no geometry, so the host can explain itself. */
    var onNoGeometry: (() -> Unit)? = null

    private val fill = Paint(Paint.ANTI_ALIAS_FLAG).apply { style = Paint.Style.FILL }
    private val stroke = Paint(Paint.ANTI_ALIAS_FLAG).apply { style = Paint.Style.STROKE }
    private val text = Paint(Paint.ANTI_ALIAS_FLAG)

    private val gestures = GestureDetector(context, object : GestureDetector.SimpleOnGestureListener() {
        override fun onDown(event: MotionEvent) = true

        override fun onFling(
            down: MotionEvent?, up: MotionEvent, vx: Float, vy: Float,
        ): Boolean {
            val dx = up.x - (down?.x ?: return false)
            // Horizontal only, and past a threshold, so a scroll or a stray
            // touch does not turn the page.
            if (abs(dx) > 120 && abs(dx) > abs(up.y - down.y)) {
                onSwipePage?.invoke(if (dx < 0) 1 else -1)
                return true
            }
            return false
        }
    })

    @Suppress("ClickableViewAccessibility") // Handled by the gesture detector.
    override fun onTouchEvent(event: MotionEvent): Boolean =
        gestures.onTouchEvent(event) || super.onTouchEvent(event)

    /**
     * Decoded masks, kept so a redraw does not decode the same letter again.
     *
     * The Android recording carries every mask's pixels on every frame — the
     * native side builds a fresh painter each time and remembers nothing — so
     * this may be cleared whenever it likes. If that ever changes, and the
     * sender starts omitting pixels for masks it believes are held here, then
     * clearing this without telling it will make text silently stop appearing,
     * exactly as the browser shell's comment warns.
     */
    private val glyphTiles = HashMap<String, Bitmap>()

    /**
     * Decode a mask, or return one already decoded.
     *
     * Guarded at every step: a recording that does not carry what this needs
     * falls back to the placeholder frame rather than throwing out of a draw.
     */
    private fun glyphTile(op: JSONObject): Bitmap? {
        val key = op.optString("key")
        if (key.isEmpty()) return null
        glyphTiles[key]?.let { return it }
        val encoded = op.optString("pixels")
        if (encoded.isEmpty()) return null
        val w = op.optInt("iw")
        val h = op.optInt("ih")
        if (w <= 0 || h <= 0) return null
        val bytes = try {
            Base64.decode(encoded, Base64.DEFAULT)
        } catch (_: IllegalArgumentException) {
            return null
        }
        if (bytes.size < w * h * 4) return null
        // The recording is one byte each of R, G, B, A; ARGB_8888 wants them
        // packed into an int the other way round.
        val pixels = IntArray(w * h)
        for (i in 0 until w * h) {
            val r = bytes[i * 4].toInt() and 0xFF
            val g = bytes[i * 4 + 1].toInt() and 0xFF
            val b = bytes[i * 4 + 2].toInt() and 0xFF
            val a = bytes[i * 4 + 3].toInt() and 0xFF
            pixels[i] = (a shl 24) or (r shl 16) or (g shl 8) or b
        }
        val tile = Bitmap.createBitmap(pixels, w, h, Bitmap.Config.ARGB_8888)
        // Every zoom level makes its own masks, so an afternoon of reading
        // would otherwise accumulate them without limit.
        if (glyphTiles.size >= MAX_GLYPH_TILES) {
            glyphTiles.clear()
        }
        glyphTiles[key] = tile
        return tile
    }

    override fun onDraw(canvas: Canvas) {
        super.onDraw(canvas)
        if (width == 0 || height == 0) return

        val ops: JSONArray = try {
            Reader.renderPage(width.toFloat(), height.toFloat(), zoom)
        } catch (_: Exception) {
            return
        }

        // Balance clips explicitly: the recording pushes and pops, and an
        // unbalanced restore here would corrupt the canvas for the whole view
        // hierarchy, not just this page.
        var clipDepth = 0

        for (index in 0 until ops.length()) {
            val op = ops.optJSONObject(index) ?: continue
            when (op.optString("op")) {
                "fill_rect" -> {
                    fill.color = parseColor(op.optString("color"))
                    canvas.drawRect(
                        op.f("x"), op.f("y"),
                        op.f("x") + op.f("w"), op.f("y") + op.f("h"), fill,
                    )
                }
                "image" -> {
                    // Every glyph on the page arrives here as a mask, so this
                    // is the text path: without it a page reads as a field of
                    // empty rectangles, which is what this drew before.
                    val tile = glyphTile(op)
                    if (tile != null) {
                        canvas.drawBitmap(
                            tile,
                            null,
                            RectF(
                                op.f("x"), op.f("y"),
                                op.f("x") + op.f("w"), op.f("y") + op.f("h"),
                            ),
                            null,
                        )
                    } else {
                        // A mask we were never sent still shows its frame:
                        // silently omitting content from a document someone is
                        // reading is worse than showing a placeholder.
                        stroke.color = Color.GRAY
                        stroke.strokeWidth = 1f
                        canvas.drawRect(
                            op.f("x"), op.f("y"),
                            op.f("x") + op.f("w"), op.f("y") + op.f("h"), stroke,
                        )
                    }
                }
                "stroke_rect" -> {
                    stroke.color = parseColor(op.optString("color"))
                    stroke.strokeWidth = op.optDouble("width", 1.0).toFloat()
                    canvas.drawRect(
                        op.f("x"), op.f("y"),
                        op.f("x") + op.f("w"), op.f("y") + op.f("h"), stroke,
                    )
                }
                "fill_path", "stroke_path" -> {
                    val points = op.optJSONArray("points") ?: continue
                    if (points.length() < 2) continue
                    val path = Path()
                    for (p in 0 until points.length()) {
                        val point = points.optJSONArray(p) ?: continue
                        val x = point.optDouble(0).toFloat()
                        val y = point.optDouble(1).toFloat()
                        if (p == 0) path.moveTo(x, y) else path.lineTo(x, y)
                    }
                    if (op.optString("op") == "fill_path") {
                        path.close()
                        fill.color = parseColor(op.optString("color"))
                        canvas.drawPath(path, fill)
                    } else {
                        stroke.color = parseColor(op.optString("color"))
                        stroke.strokeWidth = op.optDouble("width", 1.0).toFloat()
                        canvas.drawPath(path, stroke)
                    }
                }
                "line" -> {
                    stroke.color = parseColor(op.optString("color"))
                    stroke.strokeWidth = op.optDouble("width", 1.0).toFloat()
                    canvas.drawLine(op.f("x1"), op.f("y1"), op.f("x2"), op.f("y2"), stroke)
                }
                "fill_circle" -> {
                    fill.color = parseColor(op.optString("color"))
                    canvas.drawCircle(op.f("x"), op.f("y"), op.f("r"), fill)
                }
                "stroke_circle" -> {
                    stroke.color = parseColor(op.optString("color"))
                    stroke.strokeWidth = op.optDouble("width", 1.0).toFloat()
                    canvas.drawCircle(op.f("x"), op.f("y"), op.f("r"), stroke)
                }
                "text" -> {
                    text.color = parseColor(op.optString("color"))
                    text.textSize = op.f("size")
                    // The recording anchors text at its top-left to match the
                    // desktop painter; Android draws from the baseline, so the
                    // ascent is added back or every line rides too high.
                    canvas.drawText(
                        op.optString("text"),
                        op.f("x"), op.f("y") - text.fontMetrics.ascent, text,
                    )
                }
                "push_clip" -> {
                    canvas.save()
                    canvas.clipRect(
                        op.f("x"), op.f("y"),
                        op.f("x") + op.f("w"), op.f("y") + op.f("h"),
                    )
                    clipDepth++
                }
                "pop_clip" -> if (clipDepth > 0) { canvas.restore(); clipDepth-- }
                "no_geometry" -> onNoGeometry?.invoke()
            }
        }
        repeat(clipDepth) { canvas.restore() }
    }

    private fun org.json.JSONObject.f(key: String): Float = optDouble(key, 0.0).toFloat()

    /** Parse `rgba(r,g,b,a)` as the recording writes it. */
    private fun parseColor(value: String): Int {
        val parts = value.removePrefix("rgba(").removeSuffix(")").split(",")
        if (parts.size < 4) return Color.BLACK
        return try {
            Color.argb(
                ((parts[3].trim().toFloat()) * 255f).toInt().coerceIn(0, 255),
                parts[0].trim().toInt().coerceIn(0, 255),
                parts[1].trim().toInt().coerceIn(0, 255),
                parts[2].trim().toInt().coerceIn(0, 255),
            )
        } catch (_: NumberFormatException) {
            Color.BLACK
        }
    }
}

File: irondocuments-rs/apps/reader/ios/Reader/PageView.swift
// SPDX-License-Identifier: AGPL-3.0-or-later
import UIKit

/// Draws a page by replaying the recording the Rust core produced.
///
/// The page-painting decisions — glyph positions, which paths fill, how the
/// page is fitted and the y axis flipped — were all made in shared Rust. What
/// remains is a switch over primitives, which is why this file is short and why
/// iOS cannot drift from the desktop, web and Android renderings.
final class PageView: UIView {

    var zoom: CGFloat = 1 { didSet { setNeedsDisplay() } }

    /// Called when a horizontal swipe should turn the page.
    var onSwipePage: ((Int) -> Void)?

    /// Called when the document carries no geometry, so the host can say so.
    var onNoGeometry: (() -> Void)?

    override init(frame: CGRect) {
        super.init(frame: frame)
        backgroundColor = .systemBackground

        for (direction, delta) in [(UISwipeGestureRecognizer.Direction.left, 1),
                                   (.right, -1)] {
            let swipe = UISwipeGestureRecognizer(target: self, action: #selector(handleSwipe(_:)))
            swipe.direction = direction
            addGestureRecognizer(swipe)
            swipeDeltas[direction.rawValue] = delta
        }
    }

    required init?(coder: NSCoder) { fatalError("init(coder:) is not used") }

    private var swipeDeltas: [UInt: Int] = [:]

    @objc private func handleSwipe(_ gesture: UISwipeGestureRecognizer) {
        if let delta = swipeDeltas[gesture.direction.rawValue] { onSwipePage?(delta) }
    }

    /// Decoded masks, kept so a redraw does not decode the same letter again.
    ///
    /// The iOS recording carries every mask's pixels on every frame -- the
    /// native side builds a fresh painter each time and remembers nothing -- so
    /// this may be cleared whenever it likes. If that ever changes, and the
    /// sender starts omitting pixels for masks it believes are held here, then
    /// clearing this without telling it will make text silently stop appearing.
    private var glyphTiles: [String: CGImage] = [:]

    /// How many decoded masks to hold before starting again. Every zoom level
    /// makes its own, so this cannot grow without limit.
    private static let maxGlyphTiles = 4096

    /// Decode a mask, or return one already decoded.
    ///
    /// Guarded at every step: a recording that does not carry what this needs
    /// falls back to the placeholder frame rather than trapping inside a draw.
    private func glyphTile(_ op: [String: Any]) -> CGImage? {
        guard let key = op["key"] as? String, !key.isEmpty else { return nil }
        if let cached = glyphTiles[key] { return cached }
        guard let encoded = op["pixels"] as? String,
              let width = number(op["iw"]).map({ Int($0) }),
              let height = number(op["ih"]).map({ Int($0) }),
              width > 0, height > 0,
              let bytes = Data(base64Encoded: encoded),
              bytes.count >= width * height * 4,
              let provider = CGDataProvider(data: bytes as CFData)
        else { return nil }

        // The recording is straight (not premultiplied) RGBA, one byte each.
        guard let image = CGImage(
            width: width,
            height: height,
            bitsPerComponent: 8,
            bitsPerPixel: 32,
            bytesPerRow: width * 4,
            space: CGColorSpaceCreateDeviceRGB(),
            bitmapInfo: CGBitmapInfo(rawValue: CGImageAlphaInfo.last.rawValue),
            provider: provider,
            decode: nil,
            shouldInterpolate: true,
            intent: .defaultIntent
        ) else { return nil }

        if glyphTiles.count >= Self.maxGlyphTiles { glyphTiles.removeAll() }
        glyphTiles[key] = image
        return image
    }

    override func draw(_ rect: CGRect) {
        guard let context = UIGraphicsGetCurrentContext(), bounds.width > 0 else { return }

        let ops = Reader.renderPage(width: bounds.width, height: bounds.height, zoom: zoom)

        // Balance clips explicitly. The recording pushes and pops; an
        // unbalanced restore would corrupt the graphics state for everything
        // drawn after this view.
        var clipDepth = 0

        for op in ops {
            switch op["op"] as? String {
            case "fill_rect":
                context.setFillColor(color(op["color"]))
                context.fill(rect(op))
            case "image":
                // Every glyph on the page arrives here as a mask, so this is
                // the text path: without it a page reads as a field of empty
                // rectangles, which is what this drew before.
                if let tile = glyphTile(op) {
                    // Core Graphics draws images up the y axis; the recording
                    // is in screen coordinates, so flip about the destination.
                    context.saveGState()
                    let box = rect(op)
                    context.translateBy(x: 0, y: box.maxY + box.minY)
                    context.scaleBy(x: 1, y: -1)
                    context.draw(tile, in: box)
                    context.restoreGState()
                } else {
                    // A mask we were never sent still shows its frame: quietly
                    // omitting content from a document someone is reading is
                    // worse than a placeholder.
                    context.setStrokeColor(UIColor.gray.cgColor)
                    context.setLineWidth(1)
                    context.stroke(rect(op))
                }
            case "stroke_rect":
                context.setStrokeColor(color(op["color"]))
                context.setLineWidth(number(op["width"]) ?? 1)
                context.stroke(rect(op))
            case "fill_path", "stroke_path":
                guard let points = op["points"] as? [[Double]], points.count >= 2 else { break }
                context.beginPath()
                context.move(to: CGPoint(x: points[0][0], y: points[0][1]))
                for point in points.dropFirst() {
                    context.addLine(to: CGPoint(x: point[0], y: point[1]))
                }
                if op["op"] as? String == "fill_path" {
                    context.closePath()
                    context.setFillColor(color(op["color"]))
                    context.fillPath()
                } else {
                    context.setStrokeColor(color(op["color"]))
                    context.setLineWidth(number(op["width"]) ?? 1)
                    context.strokePath()
                }
            case "line":
                context.beginPath()
                context.move(to: CGPoint(x: number(op["x1"]) ?? 0, y: number(op["y1"]) ?? 0))
                context.addLine(to: CGPoint(x: number(op["x2"]) ?? 0, y: number(op["y2"]) ?? 0))
                context.setStrokeColor(color(op["color"]))
                context.setLineWidth(number(op["width"]) ?? 1)
                context.strokePath()
            case "fill_circle", "stroke_circle":
                let radius = number(op["r"]) ?? 0
                let box = CGRect(x: (number(op["x"]) ?? 0) - radius,
                                 y: (number(op["y"]) ?? 0) - radius,
                                 width: radius * 2, height: radius * 2)
                if op["op"] as? String == "fill_circle" {
                    context.setFillColor(color(op["color"]))
                    context.fillEllipse(in: box)
                } else {
                    context.setStrokeColor(color(op["color"]))
                    context.setLineWidth(number(op["width"]) ?? 1)
                    context.strokeEllipse(in: box)
                }
            case "text":
                let size = number(op["size"]) ?? 12
                let attributes: [NSAttributedString.Key: Any] = [
                    .font: UIFont.systemFont(ofSize: size),
                    .foregroundColor: UIColor(cgColor: color(op["color"])),
                ]
                // The recording anchors text at its top-left, matching every
                // other shell, and UIKit draws from the top-left too — so no
                // baseline adjustment here, unlike Android and Canvas2D.
                (op["text"] as? String ?? "").draw(
                    at: CGPoint(x: number(op["x"]) ?? 0, y: number(op["y"]) ?? 0),
                    withAttributes: attributes)
            case "push_clip":
                context.saveGState()
                context.clip(to: rect(op))
                clipDepth += 1
            case "pop_clip":
                if clipDepth > 0 { context.restoreGState(); clipDepth -= 1 }
            case "no_geometry":
                onNoGeometry?()
            default:
                break
            }
        }
        for _ in 0..<clipDepth { context.restoreGState() }
    }

    private func number(_ value: Any?) -> CGFloat? {
        (value as? NSNumber).map { CGFloat($0.doubleValue) }
    }

    private func rect(_ op: [String: Any]) -> CGRect {
        CGRect(x: number(op["x"]) ?? 0, y: number(op["y"]) ?? 0,
               width: number(op["w"]) ?? 0, height: number(op["h"]) ?? 0)
    }

    /// Parse `rgba(r,g,b,a)` as the recording writes it.
    private func color(_ value: Any?) -> CGColor {
        guard let text = value as? String,
              text.hasPrefix("rgba(") else { return UIColor.black.cgColor }
        let parts = text.dropFirst(5).dropLast()
            .split(separator: ",")
            .map { Double($0.trimmingCharacters(in: .whitespaces)) ?? 0 }
        guard parts.count >= 4 else { return UIColor.black.cgColor }
        return UIColor(red: parts[0] / 255, green: parts[1] / 255,
                       blue: parts[2] / 255, alpha: parts[3]).cgColor
    }
}
